tests/test_header_matching.js

```diff
@@ -82,7 +82,7 @@ test('match headers on number with regexp', async t => {
 })
 
 test('match header on scope with function: gets the expected argument', async t => {
-  t.plan(3)
+  t.plan(4) // The check in matchHeader should run twice
```

Comment:
Hmm, is there a reason it should? It seems confusing for the check to run more than once per request.

Reply:
We run the checks against header filters on the scope and on the interceptor
https://github.com/nock/nock/blob/beta/lib/interceptor.js#L241-L242
```
  if (
    !this.scope.matchHeaders.every(checkHeaders) ||
    !this.interceptorMatchHeaders.every(checkHeaders)
  ) {
    this.scope.logger("headers don't match")
    return false
  }
```
So I would expect:
92 in this file to trigger twice
101,102 to trigger once
For a total for 4 calls to t.equal.

Reply:
We have to check the filter against the scope filters and the interceptor filters.
```
  if (
    !this.scope.matchHeaders.every(checkHeaders) ||
    !this.interceptorMatchHeaders.every(checkHeaders)
  ) {
    this.scope.logger("headers don't match")
    return false
  }
```
So I would expect:
2 executions of line 92
1 execution of line 101
1 execution of line 102
-------------------------
4 total calls to `t.equal`

Reply:
I ran the debug and what is happening is it is called once from `matchIndependentOfBody` and once from `match`. Previously the two functions used different header matching logic. Now they use the same logic so line 92 will be called twice.

Reply:
Okay. I can dig into it a bit more and see if I can get all the tests passing with the function only being called once. Ideally I'd like to fix that before merging because I don't think there's a good reason to evaluate it twice and it seems confusing for the user.

Reply:
So `matchIndependantOfBody` is used here https://github.com/nock/nock/blob/beta/lib/intercept.js#L398 to detect if we should let the request through. But if all the check pass on `matchIndependantOfBody` we don't really do anything with it because we wait for the body when we execute match. I think what we would basically need to do to get down to 3 executions rather than 4 is to drop the whole `matchIndependantOfBody` line of execution and only executing matching once. Inside the `match` method. This would simplify nock but it looked like a lot more than I was really willing to bite off.

Reply:
Gotcha. Thank you so much for looking into it this far! I can pick it up from here.

Since the headers aren't part of the body, it seems like the check belongs in `matchIndependentOfBody`, though maybe doing it just once inside of `match` will be a more reasonable incremental change.

Reply:
You can't only check in the first because j there might be I rejecting body filter. So you could only check in the second method when you have all the data. As an alternative you could store the results of matchIndependant... on the object and then use the result in the match call rather then rerunning all the checks.

As a contributor I do find the separate matches confusing so executing the checks once would read better, but, it was so odd I assumed it was done for good reason.

Reply:
If possible you could also consider merging this change in as is and addressing the extra call separately. We would trade the duplicate call for working headed matching.